### PR TITLE
Write QA Test for Subsampling

### PR DESCRIPTION
**Pull request by knapii-developments**
Before this PR, there were no test for subsampling. Now critical functionality is tested such as:

- Groups that have numeric values are being treated as strings
- Numeric columns are rounded to 3 decimals points
- Binning correctly
- Subsampling correctly
- Filtering cell names (given from cluster file) in metadata correctly

In order to test the above cases the file requirements are:
- Has 3D data
- Has both numeric and group annotations
- Cluster file has less cells than metadata file
- Has a group annotation column that has numeric values

This PR satisfies [SCP-1835](https://broadworkbench.atlassian.net/browse/SCP-1835)

### CODE REVIEW

File: tests/test_ingest_files.py [hunk 1]
```diff
@@ -0,0 +1,95 @@
...
+
+These tests verifies:
+    - Group type annotations that have numeric-like values are being treated as strings
+    - Numeric columns are rounded to 3 decimals points
+    - Filtering cell names (given from cluster file) in metadata correctly
+
+PREREQUISITES
+Spin up Python 3.6 virtualenv, install Python dependencies in requirements.txt
+and Firestore emulator must be running, see PR26 for instructions
+(https://github.com/broadinstitute/scp-ingest-pipeline/pull/26)
```
Comment (eweitz) [inline tests/test_ingest_files.py]: Is a running Firestore emulator truly a prerequisite here?  These tests seem independent from Firestore.

File: tests/test_subsample.py [hunk 2]
```diff
@@ -0,0 +1,97 @@
+"""Tests subsampling
+
+These tests verifies:
+    - Binning correctly
+    - Subsampling correctly
+
+PREREQUISITES
+Spin up Python 3.6 virtualenv, install Python dependencies in requirements.txt
+and Firestore emulator must be running, see PR26 for instructions
+(https://github.com/broadinstitute/scp-ingest-pipeline/pull/26)
```
Comment (eweitz) [inline tests/test_subsample.py]: As in `test_ingest_files`, I suspect Firestore might not be required to run the tests in this file.

File: tests/test_subsample.py [hunk 3]
```diff
@@ -0,0 +1,97 @@
...
+import numpy as np
+
+sys.path.append("../ingest")
+
+from subsample import SubSample
+
+
+class TestSubsample(unittest.TestCase):
+    AMOUNT_OF_NUMERIC_BINS = 20
+    CLUSTER_PATH = '../tests/data/test_1k_cluster_Data.csv'
```
Comment (eweitz) [inline tests/test_subsample.py]: Conventional casing:

```suggestion
    CLUSTER_PATH = '../tests/data/test_1k_cluster_data.csv'
```

File: tests/test_subsample.py [hunk 4]
```diff
@@ -0,0 +1,97 @@
+"""Tests subsampling
+
+These tests verifies:
```
Comment (eweitz) [inline tests/test_subsample.py]:
```suggestion
These tests verify:
```

File: tests/test_ingest_files.py [hunk 5]
```diff
@@ -0,0 +1,95 @@
+"""Test ingest_files.py
+
+These tests verifies:
```
Comment (eweitz) [inline tests/test_ingest_files.py]:
```suggestion
These tests verify:
```

File: tests/test_ingest_files.py [hunk 6]
```diff
@@ -0,0 +1,95 @@
...
+PREREQUISITES
+Spin up Python 3.6 virtualenv, install Python dependencies in requirements.txt
+and Firestore emulator must be running, see PR26 for instructions
+(https://github.com/broadinstitute/scp-ingest-pipeline/pull/26)
+
+Note: When CI environment moves to Python 3.7, tests may break due to minor
+differences in how the reference issues are serialized
+
+# Run all tests in a manner that shows report_issues output
+python3 test_ingest_files.py -s
```
Comment (eweitz) [inline tests/test_ingest_files.py]: Nice refinement!

File: ingest/ingest_pipeline.py [hunk 7]
```diff
@@ -49,7 +49,11 @@
 from mtx import Mtx
 from subsample import SubSample
 from loom import Loom
-from validation.validate_metadata import validate_input_metadata, report_issues
```
Comment (jlchang) [inline ingest/ingest_pipeline.py]: validate_input_metadata will call all functions needed for a valid metadata file (including collect_jsonschema_errors and validate_collected_ontology_data). If you only call the two functions and not validate_input_metadata, then full metadata file validation is not being performed. Are there  issues with the functions recently added to validate_input_metadata (review_metadata_names and confirm_uniform_units) that need discussion?

Comment (knapii-developments) [reply ingest/ingest_pipeline.py]: I think this comment was misplaced and may have to deal with the metadata review and not subsampling. @jlchang is that correct?

File: tests/test_ingest_files.py (hunk 1 above)
Comment (knapii-developments) [reply tests/test_ingest_files.py]: You're right. Even with copy and paste, reading is essential!

File: ingest/ingest_pipeline.py (hunk 7 above)
Comment (knapii-developments) [reply ingest/ingest_pipeline.py]: I've addressed this issue in ddb16a0

File: tests/test_subsample.py (hunk 2 above)
Comment (knapii-developments) [reply tests/test_subsample.py]: 6344fd2

File: ingest/ingest_pipeline.py (hunk 7 above)
Comment (jlchang) [reply ingest/ingest_pipeline.py]: Thanks Eno. I was worried we'd silently switch back to an earlier state and not run validation in ingest_pipeline as intended. This update addresses my concerns.

**Pull request merged**